### ISSUE TITLE
Cluster Operator resets OpenShift service to remove port 9404.

### ISSUE THREAD

**Issue by Sir-Fish** [strimzi/strimzi-kafka-operator#1052]
When I try and edit the Service YAML for the Kafka Cluster to add the ports for metrics, it gets automatically removed after a brief period of time, which I assume is the cluster seeing diffs between the expected yaml and the current yaml.

------------------------------------------------------------------
What I want: 

```apiVersion: v1
kind: Service
metadata:
  creationTimestamp: '2018-10-31T19:49:19Z'
  labels:
    strimzi.io/cluster: icdc-kafka-cluster
    strimzi.io/name: icdc-kafka-cluster-kafka
    strimzi.io/type: kafka
  name: icdc-kafka-cluster-kafka
  namespace: icdc-amqstreams
  resourceVersion: '75880587'
  selfLink: /api/v1/namespaces/icdc-amqstreams/services/icdc-kafka-cluster-kafka
  uid: 0e3a155d-dd46-11e8-8157-005056a94d7b
spec:
  clusterIP: 100.124.226.175
  ports:
    - name: clients
      port: 9092
      protocol: TCP
      targetPort: 9092
    - name: replication
      port: 9091
      protocol: TCP
      targetPort: 9091
    - name: kafkametrics
      port: 9404
      protocol: TCP
      targetPort: 9404
  selector:
    strimzi.io/cluster: icdc-kafka-cluster
    strimzi.io/name: icdc-kafka-cluster-kafka
    strimzi.io/type: kafka
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```
-----------------------------------------------------------------
What it reverts to:
```
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: '2018-10-31T19:49:19Z'
  labels:
    strimzi.io/cluster: icdc-kafka-cluster
    strimzi.io/name: icdc-kafka-cluster-kafka
    strimzi.io/type: kafka
  name: icdc-kafka-cluster-kafka
  namespace: icdc-amqstreams
  resourceVersion: '75880587'
  selfLink: /api/v1/namespaces/icdc-amqstreams/services/icdc-kafka-cluster-kafka
  uid: 0e3a155d-dd46-11e8-8157-005056a94d7b
spec:
  clusterIP: 100.124.226.175
  ports:
    - name: clients
      port: 9092
      protocol: TCP
      targetPort: 9092
    - name: replication
      port: 9091
      protocol: TCP
      targetPort: 9091
  selector:
    strimzi.io/cluster: icdc-kafka-cluster
    strimzi.io/name: icdc-kafka-cluster-kafka
    strimzi.io/type: kafka
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}
```
----------------------------------------------------------

I have been attempting to figure out what is causing this, changing a bunch of different settings here and there hoping to find the fix, but I just can't seem to figure out what is going on. I apologize if this is an obvious fix, but I haven't been successful in finding any documentation other than some that say "Oh just use port 9404". 

----------------------------------------------------------

I keep thinking it is being caused by this log in the terminal on OpenShift for the ClusterController:
"
2018-10-31 19:47:24 DEBUG ServiceOperator:86 - Service icdc-amqstreams/icdc-kafka-cluster-kafka already exists, patching it
2018-10-31 19:47:24 DEBUG ServiceOperator:129 - Service icdc-kafka-cluster-kafka in namespace icdc-amqstreams has been patched
"

------------------------------------------------------------

Any help would be greatly appreciated. Feel free to ask for more clarification, and I apologize if I am not being clear, as I am fairly new to OpenShift and Strimzi.

**Comment by Sir-Fish** [comment]
I actually do have the metrics enabled in the Kafka Brokers as you can see with the image below:

-------------------------------------------------------
<img width="536" alt="kafkacluster" src="https://user-images.githubusercontent.com/38897509/47853497-4d453380-ddb5-11e8-93c7-a70f8330d97d.png">

-------------------------------------------------------

I also know I have Prometheus and Grafana set up properly, because in that short window of time where the code hasn't been corrected, the metrics are sent and can be seen both in Prometheus and Grafana. Also, the version of Strimzi I am using is not the most up to date, I believe it is about 4 months old, but I am not exactly sure which version of Strimzi we are currently using. I guess to be more specific, my question is how can I change the settings so that the service is exposed as well as the Kafka brokers so that the automatic fix doesn't remove port 9404.

**Comment by scholzj** [comment]
The version of strimzi you are using should be in the tag name of the Cluster Operator Docker image. If you tell me the version, I can give it a try.

In general, as I said - when you enable the metrics in Kafka / Zookeeper, the port 9404 should be automatically added to the service. Maybe there is some bug in your version.

**Comment by dennis-ups** [comment]
@scholzj, we are using amqstreams10-clusteroperator-openshift:latest, so I guess you are looking for the answer 1.0.0-4.

**Comment by benhwebster** [comment]
I believe I ran into this issue, or at least a similar one, but in my case it was due to the fact that there are now network policies selecting strimzi pods (since 0.7.0?), but they don't include the metrics port, which causes them to ignore traffic on that port. To get around it I created a network policy that allows access to 9404 on those pods (since you can't edit the ones created by the operator). I thought this might have been a problem with our deployment of Openshift or a configuration mistake on my end but perhaps not.

**Comment by Sir-Fish** [comment]
@scholzj - The version we are using is actually 0.2.0 for Strimzi. If you look at the image in one of the above comments that is how I have the metrics enabled, and the cluster shows that 9404 has the port sending metrics, but the service is not allowing port 9404 to be used, so we can't get that out to Prometheus. Since we are using an older version, namely 0.2.0, what would I need to change on that version to enable the metrics and service outside of the configMap and the environment variables on the cluster?

@benhwebster - I do not have cluster-admin access, but I originally thought our Strimzi version was 0.7.0, but it is actually 0.2.0, so I don't know if the network policies issue would apply, but I can still try, I just need to contact a cluster-admin to test it for me.

**Comment by scholzj** [comment]
Ok, too many people talking and that is getting a bit confusing ;-)

---

@benhwebster Yes, that was a problem. It should be fixed in 0.8.1, 0.8.2 and in master. If you don't want to upgrade (or cannot upgrade), you can just add your own network policy allowing the access to the port 9404 (just give it a custom name so that the cluster operator doesn't overwrite it). That should workaround the issue on 0.7.0 or 0.8.0.

---

@Sir-Fish Ok. 0.2.0 is really old. And looking at it we indeed didn't added the 9404 port to the services when the metrics are enabled. I guess there are two options you have. You can upgrade to some newer version. Or you can try a workaround - create you own service with your own name which would expose port 9404 and point it to the Kafka / Zookeeper pods. Since it will have a custom name and no Strimzi annotations, the Cluster operator will let it be. 

Now the screenshot is really confusing. It suggests the image is part of AMQ Streams 1.0. But the configuration is not what I would expect with AMQ Streams 1.0 GA which is based on Strimzi 0.8.1 and would look a bit different. Is it maybe the AMQ Streams Developer Preview? That was based on Strimzi 0.3.1 and might look similar.

---

@dennis-ups In AMQ Streams 1.0 GA, it should work as I described. When the metrics are enabled in the custom resource configuring the Kafka cluster, the 9404 port is automatically added to the `<cluster-name>-zookeeper-clients` and `<cluster-name>-kafka-bootstrap` services (I just double check that it actually works).

**Comment by Sir-Fish** [comment]
@scholzj - We are using the AMQStreams Developer Preview. I was under the impression that the Preview was 0.2.0 and not 0.3.1, sorry if that confused you. But I will try setting up a separate service and see how that goes. We are already working on upgrading to a newer version of Strimzi, so hopefully that will be up and running soon. Thank you for your help.
